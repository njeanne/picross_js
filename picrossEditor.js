/* eslint-disable no-plusplus */
// la classe
export class PicrossEditor {
  constructor(node) {
    this.name = 'picross editor';
    this.node = node;
  }

  // générer le tableau de solution
  static getCellStateByOrientation(rows, useFirstRowAndFirstCol = true) {
    const cellStateByRow = [];
    const cellStateByCol = [];
    rows.forEach((currentTr, idxRow) => {
      cellStateByRow[idxRow] = [];
      let allTd;
      if (useFirstRowAndFirstCol) {
        allTd = currentTr.querySelectorAll('td');
      } else {
        allTd = currentTr.querySelectorAll('td:not(td:first-of-type)');
      }
      allTd.forEach((currentTd, idxCol) => {
        // utilisation du nullish coalescing assignment ??= qui vérifie
        // si le tableau existe, si il existe rien ne se passe sinon il créé le tableau
        cellStateByCol[idxCol] ??= [];
        if (currentTd.classList.contains('bg-blue')) {
          cellStateByRow[idxRow][idxCol] = true;
          cellStateByCol[idxCol][idxRow] = true;
        } else {
          cellStateByRow[idxRow][idxCol] = false;
          cellStateByCol[idxCol][idxRow] = false;
        }
      });
    });
    return { rowStates: cellStateByRow, colStates: cellStateByCol };
  }

  // extrait le nombre de cellules occupées par ligne ou colonne
  static extractUseCells(cellsByOrientation) {
    const indications = [];
    cellsByOrientation.forEach((currentCellsStateByOrientation) => {
      const indication = [];
      let cellsEnabledNumber = 0;
      currentCellsStateByOrientation.forEach((currentCell) => {
        if (currentCell === true) {
          cellsEnabledNumber++;
        } else {
          if (cellsEnabledNumber !== 0) {
            indication.push(cellsEnabledNumber);
          }
          cellsEnabledNumber = 0;
        }
      });
      if (cellsEnabledNumber !== 0) {
        indication.push(cellsEnabledNumber);
      }
      indications.push(indication);
    });
    return indications;
  }

  // récupérer le nombre de colonnes et de lignes et créer le tableau de jeu
  createPycross(nRow, nCol) {
    const table = document.createElement('table');
    table.id = this.node;
    for (let i = 0; i < nRow; i++) {
      const row = document.createElement('tr');
      for (let j = 0; j < nCol; j++) {
        const col = document.createElement('td');
        // switcher en fond bleu des cellules du tableau ou non des cellules si clique
        col.onclick = (ev) => {
          ev.target.classList.toggle('bg-blue');
        };
        row.appendChild(col);
      }
      table.appendChild(row);
    }
    document.getElementById(this.node).replaceWith(table);
  }

  // récupérer la solution
  getSolution(title) {
    const allTr = document.getElementById(this.node).querySelectorAll('tr');
    const states = PicrossEditor.getCellStateByOrientation(allTr);
    const solution = {
      // si 'title,' il affecte la variable title au nom title
      title,
      data: {
        rows: PicrossEditor.extractUseCells(states.rowStates),
        columns: PicrossEditor.extractUseCells(states.colStates),
      },
      size: {
        rows: states.rowStates.length,
        columns: states.colStates.length,
      },
    };
    console.log(solution);
    return solution;
  }
}
