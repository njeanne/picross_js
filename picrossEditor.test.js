import { describe, expect, test } from 'vitest';
import { PicrossEditor } from './picrossEditor';

describe('picrossEditor', () => {
  test("instanciation de l'éditeur", () => {
    const editor = new PicrossEditor('picross_player');
    expect(editor).toBeDefined();
  });

  test('création de picross', () => {
    // création de la table expected
    const table = document.createElement('table');
    table.id = 'picross_player';
    document.body.appendChild(table);
    // création de la table picross
    const editor = new PicrossEditor('picross_player');
    editor.createPycross(3, 4);
    // le test
    expect(document.getElementById('picross_player')).toMatchSnapshot();
    // on nettoie le DOM en enlevant la table expected
    document.body.removeChild(document.getElementById('picross_player'));
  });

  test('modification du picross en cliquant sur les cellules', () => {
    // création de la table expected
    const table = document.createElement('table');
    table.id = 'picross_player';
    document.body.appendChild(table);
    // création de la table picross
    const editor = new PicrossEditor('picross_player');
    editor.createPycross(3, 4);
    // on créé des cliques sur les cellules pour transformer le fond en bleu
    document.querySelector('table tr td').click();
    document.querySelector('table tr:nth-of-type(2) td:nth-of-type(3)').click();
    // le test
    expect(document.getElementById('picross_player')).toMatchSnapshot();
    // on nettoie le DOM en enlevant la table expected
    document.body.removeChild(document.getElementById('picross_player'));
  });
});
