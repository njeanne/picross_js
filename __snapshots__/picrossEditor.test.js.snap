// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`picrossEditor > cr√©ation de picross 1`] = `
<table
  id="picross_player"
>
  <tr>
    <td />
    <td />
    <td />
    <td />
  </tr>
  <tr>
    <td />
    <td />
    <td />
    <td />
  </tr>
  <tr>
    <td />
    <td />
    <td />
    <td />
  </tr>
</table>
`;

exports[`picrossEditor > modification du picross en cliquant sur les cellules 1`] = `
<table
  id="picross_player"
>
  <tr>
    <td
      class="bg-blue"
    />
    <td />
    <td />
    <td />
  </tr>
  <tr>
    <td />
    <td />
    <td
      class="bg-blue"
    />
    <td />
  </tr>
  <tr>
    <td />
    <td />
    <td />
    <td />
  </tr>
</table>
`;
